symforce/slam/imu_preintegration/imu_preintegrator.h

```diff
@@ -0,0 +1,78 @@
+/* ----------------------------------------------------------------------------
+ * SymForce - Copyright 2022, Skydio, Inc.
+ * This source code is under the Apache 2.0 license found in the LICENSE file.
+ * ---------------------------------------------------------------------------- */
+
+#pragma once
+
+#include <Eigen/Core>
+
+#include <sym/util/epsilon.h>
+
+#include "./preintegrated_imu_measurements.h"
+
+namespace sym {
+
+/**
+ * Class to on-manifold preintegrate IMU measurements for usage in a SymForce optimization
+ * problem.
+ *
+ * For usage, see the doc-string of ImuFactor in imu_factor.h
+ */
+template <typename Scalar>
+class ImuPreintegrator {
+ public:
+  using Vector3 = typename PreintegratedImuMeasurements<Scalar>::Vector3;
+  using Matrix33 = typename PreintegratedImuMeasurements<Scalar>::Matrix33;
+  using Matrix99 = Eigen::Matrix<Scalar, 9, 9>;
+
+ private:
+  PreintegratedImuMeasurements<Scalar> preintegrated_measurments_;
+  Matrix99 covariance_;  // covariance of [DR.ToTangent(), Dv, Dp]
+
+ public:
+  /**
+   * Initialize with given accel_bias and gyro_bias
+   */
+  ImuPreintegrator(const Vector3& accel_bias, const Vector3& gyro_bias);
+
+  /**
+   * Integrate a new measurement.
+   *
+   * Precondition:
+   *   *this was initialized by a constructor and has only been mutated by IntegrateMeasurement
+   *     (called in order on consecutive IMU measurements)
+   *   measured_accel is the next accelerometer measurement after the last integrated measurement
+   *   measured_gyro is the next gyroscope measurement after the last integrated measurement
+   *   accel_cov is the covariance of the accelerometer measurement
+   *   gyro_cov is the covariance of the gyroscope measurement
```

Comment:
Also mentioned [here](https://github.com/symforce-org/symforce/pull/280#discussion_r1073003081), but we should say what units there are in - assuming the test is right it seems to me like maybe these aren't technically the covariances of the measurements, they're the imu noise densities?

Reply:
Yeah, the tests are right. So the units should be `(radians per time)^2 * time` (or, perhaps more recognizably to some, `(radians per time)^2 / frequency`).

And I suppose it is also important to mention that the units are expected to be in terms of radians, not degrees.

I don't know of any simple words that clearly communicate the idea that it's the, idk, covariance density in frequency space or whatever, so I'll just append `[(radians / time)^2 * time]` to the ends of the lines.

Reply:
Probably good to specify seconds here for the time also, since these numbers are sometimes quoted in hours - I think the square roots of these are sometimes called angle random walk / velocity random walk, but just the units should be sufficient

Reply:
I do wonder if it's more expected / intuitive for these parameters to be the covariances, or the (squared) angle/velocity random walks - maybe @chao-qu-skydio @dominic-skydio have opinions?  Any idea what other libraries (e.g. GTSAM) expect?

Reply:
Are you talking about the names? I think xxx_cov is fine. My suggestion is to simply make it a 3x1 vector instead of 3x3 matrix. I've never seen a use case for a full covariance matrix on imu measurement.

Reply:
Talking about the values - is it more expected for these to be the measurement covariances (`(rad/s)^2`) or the squared angle/velocity random walks (`(rad / sqrt(Hz))^2`)?  I do wonder if they're the (squared) angle/velocity random walks if we should maybe name them differently, but I agree naming it xxx_cov is probably fine in either case

Reply:
About making these diagonal, it does save a good number of ops: https://github.com/symforce-org/symforce/pull/279#discussion_r1057990455

Reply:
They should be the continuous-time noise density squared. The random walk is for the bias.
Most imu datasheets only report a single number for all 3 axes (or at most 3, 1 for each), so a full covariance is an overkill IMO.

Reply:
Well, I've changed the covariances to just be the diagonal entries, as it gives an immediate 10% op-count reduction for the integration step, and it should always be easy to go back and add a full square covariance option if we really wanted it.